### PR TITLE
[libc] Allow BigInt class to use base word types other than uint64_t.

### PR DESCRIPTION
**Pull request by lntue**
This will allow DyadicFloat class to replace NormalFloat class.

### CODE REVIEW

File: libc/src/__support/UInt.h [hunk 1]
```diff
@@ -403,127 +408,139 @@ template <size_t Bits, bool Signed> struct BigInt {
...
   // performing 32-bit shiftings instead of bit-by-bit shiftings.
-  //   Since the remainder of each division step < x < 2^32, the computation of
-  // each step is now properly contained within uint64_t.
+  //   Since the remainder of each division step < x < 2^(WORD_SIZE / 2), the
+  // computation of each step is now properly contained within WordType.
   //   And finally we perform some extra alignment steps for the remaining bits.
-  LIBC_INLINE constexpr optional<BigInt<Bits, Signed>>
-  div_uint32_times_pow_2(uint32_t x, size_t e) {
-    BigInt<Bits, Signed> remainder(0);
+  // template <typename T>
```
Comment (michaelrj-google) [inline libc/src/__support/UInt.h:439]: it looks like you decided to not go with this design, should the code be deleted and comments reverted?

**Comment by gchatelet** [comment]
I haven't read the patch yet but just to make sure we're on the same page I'm cross posting this here https://github.com/llvm/llvm-project/pull/81267#issuecomment-1941301356

File: libc/src/__support/UInt.h (hunk 1 above)
Comment (lntue) [reply libc/src/__support/UInt.h]: I've removed the comments and fixed the input types.

File: libc/src/__support/UInt.h [hunk 2]
```diff
@@ -254,33 +258,33 @@ template <size_t Bits, bool Signed> struct BigInt {
...
-        uint64_t carry = 0;
-        for (size_t i = 0; i < WORDCOUNT; ++i) {
+        BigInt<Bits, Signed, WordType> result(0);
+        BigInt<2 * WORD_SIZE, Signed, WordType> partial_sum(0);
+        WordType carry = 0;
+        for (size_t i = 0; i < WORD_COUNT; ++i) {
           for (size_t j = 0; j <= i; j++) {
-            NumberPair<uint64_t> prod = full_mul(val[j], other.val[i - j]);
-            BigInt<128, Signed> tmp({prod.lo, prod.hi});
+            NumberPair<WordType> prod = full_mul(val[j], other.val[i - j]);
```
[the remark this answers is not in the document]
Comment (lntue) [reply libc/src/__support/UInt.h:299]: I've updated `full_mul` to make it works for other types, and added tests for `__uint128_t` base type.

**Comment by lntue** [comment]
> I haven't read the patch yet but just to make sure we're on the same page I'm cross posting this here [#81267 (comment)](https://github.com/llvm/llvm-project/pull/81267#issuecomment-1941301356)

Thanks, that patch would definitely help with initializing `UInt` classes.  Hopefully there won't be any issue merging this and https://github.com/llvm/llvm-project/pull/81267

**Review by michaelrj-google — APPROVED** [review]
LGTM

**Comment by gchatelet** [comment]
> > I haven't read the patch yet but just to make sure we're on the same page I'm cross posting this here [#81267 (comment)](https://github.com/llvm/llvm-project/pull/81267#issuecomment-1941301356)
> 
> Thanks, that patch would definitely help with initializing `UInt` classes. Hopefully there won't be any issue merging this and #81267

I synced #81267 and unfortunately there is one issue
```
libc/src/__support/integer_literals.h:122:28: error: class template partial specialization contains a template parameter that cannot be deduced; this partial specialization will never be used [-Wunusable-partial-specialization]
template <size_t N> struct Parser<LIBC_NAMESPACE::cpp::UInt<N>> {
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
libc/src/__support/integer_literals.h:122:18: note: non-deducible template parameter 'N'
template <size_t N> struct Parser<LIBC_NAMESPACE::cpp::UInt<N>> {
```
I need to think this through.

File: libc/test/src/__support/uint_test.cpp [hunk 3]
```diff
@@ -676,6 +676,52 @@ TEST(LlvmLibcUIntClassTest, ConstructorFromUInt128Tests) {
...
+
+  ASSERT_EQ(static_cast<int>(a), 1);
+  a = (a << 128) + 2;
+  ASSERT_EQ(static_cast<int>(a), 2);
+  ASSERT_EQ(static_cast<uint64_t>(a), uint64_t(2));
+  a = (a << 32) + 3;
+  ASSERT_EQ(static_cast<int>(a), 3);
+  ASSERT_EQ(static_cast<uint64_t>(a), uint64_t(0x2'0000'0003));
+  ASSERT_EQ(static_cast<int>(a >> 32), 2);
+  ASSERT_EQ(static_cast<int>(a >> (128 + 32)), 1);
```
Comment (gchatelet) [inline libc/test/src/__support/uint_test.cpp:693]: Shouldn't this be `ASSERT_EQ(static_cast<int>(a >> (128 + 32)), 0);`?
This line started failing when running tests for #86137 .

Comment (lntue) [reply libc/test/src/__support/uint_test.cpp:693]: No, after L689, the value of `a` is `2^(128+32) + 2 × 2^32 + 3`, so right shift by `128 + 32` should make it 1.

Comment (gchatelet) [reply libc/test/src/__support/uint_test.cpp:693]: It was a bug in my code :) Sorry for the noise.

**Pull request merged**